### DISCUSSION TITLE
Lazy load options (AsyncSelect)

### DISCUSSION THREAD

**Discussion by feyst** [JedWatson/react-select#5928] [Ideas]
Hello,

On our system we have many select fields causing many requests on the page when the page is opened. The selects are used as filters so most of them are not used on average. I couldn't find any documentation about loading the options only after clicking inside / start searching in the select.

I could only find a closed issue: [https://github.com/JedWatson/react-select/issues/738](https://github.com/JedWatson/react-select/issues/738)

I can help create a PR if you think this is something you would like to support.

Kind regards,
Mark

**Comment by s-hoff** [comment]
I need the same, plus the ability to re-do the search every time the select is opened or the search is emptied. I already made #5986 before finding this, so I'll add a copy of that text here and close my discussion.

**Comment by s-hoff** [comment]
I want to use searchable Async Selects for a number of inputs. In them, an empty search input, both when first opening the select and emptying a previous search, should result in a search with no filter.

I realise that I can set defaultOptions to get something like this, but it only makes an empty search on mount.
I want to
  - not search before actually opening the select,
  - search every time I open the select and
  - search again when emptying the select

Especially the second part is very important in two cases:
  - The data may have changed since the first page load
  - Other inputs on the form also affect the options that should be visible

I know that it is possible to re-trigger the defaultOptions search by changing the components key, but this feels like a very hacky way of handling it and still cannot reflect possible changes in the data source.

A property like `allowEmptySearch` could be used to avoid a breaking change from this, although I would argue that this behaviour (treating empty search fields the same as filled ones) is what people would actually expect. I certainly did, and my colleagues and I were quite surprised by the current behaviour.

If a min length like #5832 proposes s introduced with a default of 1, it would also avoid the breaking change as setting it to 0 would already indicate that loads from searches are wanted.

If done without an extra property, `defaultOptions` should probably override this behaviour to avoid having the options list change shortly after opening and without user input.

I took a brief look at the code and it looks simple enough to implement, so I'd happily contribute the changes myself.

**Comment by lindenwells** [comment]
Does this meet your requirements? If you could make a promise to return your options, it might do what you need.
https://react-select.com/async

I'm looking into this myself now, I'll see if I can get something to work.
